### PR TITLE
OpenAI Files & Vector Store Hooks

### PR DESCRIPTION
**Pull request by nathadfield**
Adding a series of hook functions for OpenAI [File](https://platform.openai.com/docs/api-reference/files) and [Vector Store](https://platform.openai.com/docs/api-reference/vector-stores) endpoints.

### CODE REVIEW

File: airflow/providers/openai/hooks/openai.py [hunk 1]
```diff
@@ -133,26 +143,26 @@ def get_assistants(self, **kwargs: Any) -> list[Assistant]:
...
 
         :param assistant_name: The name of the assistant to retrieve
         """
-        response = self.get_assistants()
-        for assistant in response:
-            if assistant.name == assistant_name:
-                return assistant
-        return None
+        assistants = self.get_assistants()
+        return next((assistant for assistant in assistants if assistant.name == assistant_name), None)
```
Comment (Lee-W) [inline airflow/providers/openai/hooks/openai.py]: I'm not sure. I feel like it's a bit harder to read 🤔 Is there any reason we need to do it this way? (maybe performance?) also, I just notice this only returns the first found. is it possible to have multiple ones with the same name?

File: airflow/providers/openai/hooks/openai.py [hunk 2]
```diff
@@ -257,11 +287,137 @@ def create_embeddings(
...
+
+    def upload_file(self, file: str, purpose: Literal["fine-tune", "assistants"]) -> FileObject:
+        """
+        Upload a file that can be used across various endpoints. The size of all the files uploaded by one organization can be up to 100 GB.
+
+        :param file: The File object (not file name) to be uploaded.
+        :param purpose: The intended purpose of the uploaded file. Use "fine-tune" for
+            Fine-tuning and "assistants" for Assistants and Messages.
+        """
+        file_object = self.conn.files.create(file=open(file, "rb"), purpose=purpose)
```
Comment (Lee-W) [inline airflow/providers/openai/hooks/openai.py]: Should we use a context manger for `open`? e.g., 

```python
with open(...):
    file_object...
```

File: airflow/providers/openai/hooks/openai.py [hunk 3]
```diff
@@ -257,11 +287,137 @@ def create_embeddings(
...
+    def get_file_by_name(self, file_name: str) -> FileObject | None:
+        """
+        Get an OpenAI Assistant object for a given name.
+
+        :param file_name: The name of the file object to retrieve
+        """
+        files = self.get_files()
+        return next((file for file in files if file.filename == file_name), None)
+
+    def delete_file(self, file_id) -> FileDeleted:
```
Comment (Lee-W) [inline airflow/providers/openai/hooks/openai.py]:
```suggestion
    def delete_file(self, file_id: str) -> FileDeleted:
```

File: airflow/providers/openai/hooks/openai.py [hunk 4]
```diff
@@ -257,11 +287,137 @@ def create_embeddings(
...
+        """
+        Delete a file.
+
+        :param file_id: The ID of the file to be deleted.
+        """
+        response = self.conn.files.delete(file_id=file_id)
+        return response
+
+    def create_vector_store(self, **kwargs) -> VectorStore:
```
Comment (Lee-W) [inline airflow/providers/openai/hooks/openai.py]:
```suggestion
    def create_vector_store(self, **kwargs: Any) -> VectorStore:
```

File: airflow/providers/openai/hooks/openai.py [hunk 5]
```diff
@@ -257,11 +287,137 @@ def create_embeddings(
...
+        """Create a vector store."""
+        vector_store = self.conn.beta.vector_stores.create(**kwargs)
+        return vector_store
+
+    def get_vectors_stores(self, **kwargs) -> list[VectorStore]:
```
Comment (Lee-W) [inline airflow/providers/openai/hooks/openai.py]:
```suggestion
    def get_vectors_stores(self, **kwargs: Any) -> list[VectorStore]:
```

File: airflow/providers/openai/hooks/openai.py [hunk 6]
```diff
@@ -257,11 +287,137 @@ def create_embeddings(
...
+        Get an OpenAI Vector Store object for a given name.
+
+        :param vector_store_name: The name of the vector store to retrieve.
+        """
+        vector_stores = self.get_vectors_stores()
+        return next(
+            (vector_store for vector_store in vector_stores if vector_store.name == vector_store_name), None
+        )
+
+    def modify_vector_store(self, vector_store_id: str, **kwargs) -> VectorStore:
```
Comment (Lee-W) [inline airflow/providers/openai/hooks/openai.py]:
```suggestion
    def modify_vector_store(self, vector_store_id: str, **kwargs: Any) -> VectorStore:
```

File: airflow/providers/openai/hooks/openai.py (hunk 1 above)
Comment (nathadfield) [reply airflow/providers/openai/hooks/openai.py]: Yes, I figured this was better in terms of performance but you make a good a point.  I've not actually checked if it's possible to create two assistants with the same name.  I'll find out.

Comment (nathadfield) [reply airflow/providers/openai/hooks/openai.py]: Yes, you can, which is perhaps not unsurprising.  In that case, I think I will remove the `get_x_by_name` functions.

File: airflow/providers/openai/hooks/openai.py [hunk 7]
```diff
@@ -257,11 +278,118 @@ def create_embeddings(
...
+        """
+        response = self.conn.files.delete(file_id=file_id)
+        return response
+
+    def create_vector_store(self, **kwargs: Any) -> VectorStore:
+        """Create a vector store."""
+        vector_store = self.conn.beta.vector_stores.create(**kwargs)
+        return vector_store
+
+    def get_vectors_stores(self, **kwargs: Any) -> list[VectorStore]:
```
Comment (Lee-W) [inline airflow/providers/openai/hooks/openai.py]:
```suggestion
    def get_vector_stores(self, **kwargs: Any) -> list[VectorStore]:
```

File: tests/providers/openai/hooks/test_openai.py [hunk 8]
```diff
@@ -296,6 +397,103 @@ def test_create_embeddings(mock_openai_hook, mock_embeddings_response):
...
+def test_get_vector_store(mock_openai_hook, mock_vector_store):
+    mock_openai_hook.conn.beta.vector_stores.retrieve.return_value = mock_vector_store
+    vector_store = mock_openai_hook.get_vector_store(vector_store_id=VECTOR_STORE_ID)
+    assert vector_store.id == VECTOR_STORE_ID
+    assert vector_store.name == VECTOR_STORE_NAME
+
+
+def test_get_vector_stores(mock_openai_hook, mock_vector_store_list):
+    mock_openai_hook.conn.beta.vector_stores.list.return_value = mock_vector_store_list
+    vector_stores = mock_openai_hook.get_vectors_stores()
```
Comment (Lee-W) [inline tests/providers/openai/hooks/test_openai.py]:
```suggestion
    vector_stores = mock_openai_hook.get_vector_stores()
```

**Review by Lee-W — APPROVED** [review]
Left one nitpicks. But rest looks good to me 👍

**Review by Lee-W — APPROVED** [review]

**Comment by nathadfield** [comment]
Any chance of a merge before the next round of providers are cut?

**Comment by Lee-W** [comment]
> Any chance of a merge before the next round of providers are cut?

The next wave has already started. We'll probably need to wait for the next next wave, but I can merge it now.

**Comment by nathadfield** [comment]
@Lee-W Thanks!  No worries.

**Pull request merged**